### PR TITLE
feat: add `AlwaysDeploy` to `AWS::Serverless::Api`

### PR DESCRIPTION
**Pull request by hoffa**
### Issue #, if available

#660 

And the ones deduplicated into the above:

1. https://github.com/aws/serverless-application-model/issues/634
1. https://github.com/aws/serverless-application-model/issues/930
1. https://github.com/aws/serverless-application-model/issues/479
1. https://github.com/aws/serverless-application-model/issues/914
1. https://github.com/aws/serverless-application-model/issues/1162
1. https://github.com/aws/serverless-application-model/issues/1561
1. https://github.com/aws/serverless-application-model/issues/1447
1. https://github.com/aws/serverless-application-model/issues/2888

### Description of changes

Add new property `AlwaysDeploy` (disabled by default) to `AWS::Serverless::Api` that always creates new deployment.

Using the time instead of something deterministic (e.g. hash of template and parameters) in case something outside of SAM's visibility changes (e.g. S3 object).

### Description of how you validated changes

Based on template from https://github.com/aws/serverless-application-model/issues/2888, save as `template.yaml`:

```yaml
Transform: AWS::Serverless-2016-10-31
Resources:
  MyFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.handler
      InlineCode: |
        exports.handler = async function (event) {
          return {
            statusCode: 200,
            body: JSON.stringify({ message: "Hello, foo!" }),
          }
        }
      Runtime: nodejs16.x
      AutoPublishAlias: live

  MyRole: 
    Type: AWS::IAM::Role
    Properties:
      AssumeRolePolicyDocument: 
        Version: "2012-10-17"
        Statement: 
        - Effect: Allow
          Principal: {Service: "apigateway.amazonaws.com"}
          Action: sts:AssumeRole
      Policies: 
      - PolicyName: AllowInvokeLambdaFunctions
        PolicyDocument: 
          Version: "2012-10-17"
          Statement: 
          - Effect: Allow
            Action: lambda:InvokeFunction
            Resource: "*"

  MyApi:
    Type: AWS::Serverless::Api
    Properties:
       StageName: dev
       DefinitionBody:
        openapi: 3.0.0
        paths:
          /hello:
            get:
              x-amazon-apigateway-integration:
                credentials: !GetAtt MyRole.Arn
                uri: !Sub arn:aws:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${MyFunction.Version}/invocations
                passthroughBehavior: when_no_match
                httpMethod: POST
                type: aws_proxy

Outputs:
  WebEndpoint:
    Description: "API Gateway endpoint URL"
    Value: !Sub "https://${MyApi}.execute-api.${AWS::Region}.amazonaws.com/dev/hello"
```

Deploy:

```bash
../serverless-application-model/bin/sam-translate.py --template-file template.yaml && sam deploy --region us-west-2 --resolve-s3 --capabilities CAPABILITY_IAM --stack-name test-always-deploy --template transformed-template.yaml
```

`curl` endpoint:

```json
{"message":"Hello, foo!"}
```

Now change the Lambda:

```bash
sed -i.bak 's/Hello, foo!/Hello, bar!/g' template.yaml
```

Redeploy:

```bash
../serverless-application-model/bin/sam-translate.py --template-file template.yaml && sam deploy --region us-west-2 --resolve-s3 --capabilities CAPABILITY_IAM --stack-name test-always-deploy --template transformed-template.yaml
```

But the endpoint still shows the old output:

```json
{"message":"Hello, foo!"}
```

Now, testing the fix, change `template.yaml` as such:

```diff
--- before.yaml	2023-02-28 10:19:46
+++ template.yaml	2023-02-28 10:20:03
@@ -35,6 +35,7 @@
   MyApi:
     Type: AWS::Serverless::Api
     Properties:
+       AlwaysDeploy: true
        StageName: dev
        DefinitionBody:
         openapi: 3.0.0
```

And redeploy:

```bash
../serverless-application-model/bin/sam-translate.py --template-file template.yaml && sam deploy --region us-west-2 --resolve-s3 --capabilities CAPABILITY_IAM --stack-name test-always-deploy --template transformed-template.yaml
```

And `curl` the endpoint again (sometimes takes a few seconds to change), and we get the proper output:

```json
{"message":"Hello, bar!"}
```

### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/apigateway.py [hunk 1]
```diff
@@ -121,6 +123,9 @@ def make_auto_deployable(
             hash_input.append(str(openapi_version))
         if domain:
             hash_input.append(json.dumps(domain))
+        if always_deploy:
+            # Using int so tests are a little more robust; don't think the Python spec defines default precision
+            hash_input.append(str(int(time.time())))
```
Comment (aahung) [inline samtranslator/model/apigateway.py:128]: do we still need `swagger`, `openapi_version` and `domain` as a part of hash if `AlwaysDeploy` is true?

File: samtranslator/model/sam_resources.py [hunk 2]
```diff
@@ -1168,6 +1168,7 @@ class SamApi(SamResourceMacro):
         "Mode": PropertyType(False, IS_STR),
         "DisableExecuteApiEndpoint": PropertyType(False, is_type(bool)),
         "ApiKeySourceType": PropertyType(False, IS_STR),
+        "AlwaysDeploy": Property(False, is_type(bool)),
```
Comment (aahung) [inline samtranslator/model/sam_resources.py:1171]: nit and irrelavant: can we create a `IS_BOOL`

File: tests/translator/test_translator.py [hunk 3]
```diff
@@ -657,6 +660,93 @@ def _do_transform(self, document, parameter_values=get_template_parameter_values
         return output_fragment
 
 
+class TestApiAlwaysDeploy(TestCase):
```
Comment (aahung) [inline tests/translator/test_translator.py:663]: can we have some transform tests with `time` mocked so we know it kind of work end to end

Comment (hoffa) [reply tests/translator/test_translator.py:663]: It is there.

File: samtranslator/model/apigateway.py (hunk 1 above)
Comment (hoffa) [reply samtranslator/model/apigateway.py:128]: No, we just care that it changes. But due to how the existing code is written (chained `if`s, no separate function), doing it in another way would make the diff messy.

File: samtranslator/model/sam_resources.py (hunk 2 above)
Comment (hoffa) [reply samtranslator/model/sam_resources.py:1171]: Yes.

**Comment by xazhao** [comment]
How about `AWS::Serverless::HttpApi`?

File: samtranslator/model/apigateway.py (hunk 1 above)
Comment (hoffa) [reply samtranslator/model/apigateway.py]: Switched to assignment instead so future refactor is easier (no hash change).

**Comment by hoffa** [comment]
> How about `AWS::Serverless::HttpApi`?

I haven't seen a similar issue reported for it. `AWS::ApiGatewayV2::Stage` has [`AutoDeploy`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-stage.html#cfn-apigatewayv2-stage-autodeploy) FWIW.

File: samtranslator/model/sam_resources.py [hunk 4]
```diff
@@ -1197,6 +1198,7 @@ class SamApi(SamResourceMacro):
     Mode: Optional[Intrinsicable[str]]
     DisableExecuteApiEndpoint: Optional[Intrinsicable[bool]]
     ApiKeySourceType: Optional[Intrinsicable[str]]
+    AlwaysDeploy: Optional[bool]
```
Comment (GavinZZ) [inline samtranslator/model/sam_resources.py:1201]: Are we going to support `AlwaysDeploy` as a global?

File: tests/translator/test_translator.py [hunk 5]
```diff
@@ -657,6 +660,93 @@ def _do_transform(self, document, parameter_values=get_template_parameter_values
...
+            if v["Type"] == "AWS::ApiGateway::Deployment":
+                deployment_ids.add(k)
+        return deployment_ids
+
+    @patch("boto3.session.Session.region_name", "ap-southeast-1")
+    @patch("botocore.client.ClientEndpointBridge._check_default_region", mock_get_region)
+    def test_always_deploy(self):
+        with patch("time.time", lambda: 13.37):
+            path = "tests/translator/input/translate_always_deploy.yaml"
+            obj = yaml_parse(PROJECT_ROOT.joinpath(path).read_text())
```
Comment (GavinZZ) [inline tests/translator/test_translator.py:686]: Why not use `INPUT_FOLDER`?

Comment (hoffa) [reply tests/translator/test_translator.py:686]: I like my paths as `Path`s.

File: samtranslator/model/sam_resources.py (hunk 4 above)
Comment (hoffa) [reply samtranslator/model/sam_resources.py:1201]: I was originally thinking no since it's more of a workaround, but it doesn't hurt. I've added it.

**Review by xazhao — APPROVED** [review]

**Review by GavinZZ — APPROVED** [review]

**Review by aahung — APPROVED** [review]

**Pull request merged**